contracts/staking/ServiceStakingToken.sol

```diff
@@ -0,0 +1,199 @@
+    /// @param token Token address.
+    /// @param to Address to transfer tokens to.
+    /// @param amount Token amount.
+    function safeTransfer(address token, address to, uint256 amount) internal {
+        bool success;
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+        // We'll write our calldata to this slot below, but restore it later.
+            let memPointer := mload(0x40)
+
+        // Write the abi-encoded calldata into memory, beginning with the function selector.
+            mstore(0, 0xa9059cbb00000000000000000000000000000000000000000000000000000000)
+            mstore(4, to) // Append the "to" argument.
+            mstore(36, amount) // Append the "amount" argument.
+
+            success := and(
+            // Set success to whether the call reverted, if not we check it either
+            // returned exactly 1 (can't just be non-zero data), or had no return data.
+                or(and(eq(mload(0), 1), gt(returndatasize(), 31)), iszero(returndatasize())),
+            // We use 68 because that's the total length of our calldata (4 + 32 * 2)
+            // Counterintuitively, this call() must be positioned after the or() in the
+            // surrounding and() because and() evaluates its arguments from right to left.
+                call(gas(), token, 0, 0, 68, 0, 32)
+            )
+
+            mstore(0x60, 0) // Restore the zero slot to zero.
+            mstore(0x40, memPointer) // Restore the memPointer.
+        }
+
+        if (!success) {
+            revert TransferFailed(token, address(this), to, amount);
+        }
+    }
+
+    /// @dev Checks token security deposit.
+    /// @param serviceId Service Id.
+    function _checkTokenSecurityDeposit(uint256 serviceId) internal view override {
+        // Get the service security token and deposit
+        (address token, uint96 securityDeposit) =
+            IServiceTokenUtility(serviceRegistryTokenUtility).mapServiceIdTokenDeposit(serviceId);
+
+        // The security token must match the contract token
+        if (securityToken != token) {
+            revert();
+        }
+
+        // The security deposit must be greater or equal to the minimum defined one
+        if (securityDeposit < minSecurityDeposit) {
+            revert();
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal override {
+        safeTransfer(securityToken, to, amount);
+
+        emit Withdraw(to, amount);
+    }
+
+    /// @dev Deposits funds for staking.
+    /// @param amount Token amount to deposit.
+    function deposit(uint256 amount) external {
+        // Distribute current staking rewards
+        _checkpoint(0);
+
+        // Add to the overall balance
+        safeTransferFrom(securityToken, msg.sender, address(this), amount);
```

Comment:
Why can we not work with an interface approach? This way of doing it looks messy to me.

Reply:
We should assume an IERC20 interface https://ethereum.org/en/developers/docs/standards/tokens/erc-20/ and so the "safe" versions are not present. Will have to add appropriate checks accordingly

Reply:
See above

Reply:
A more detailed history of why it appeared "safe" ERC20:
https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca

In fact, it has been discussed for a long time to make this part of the standard:
https://eips.ethereum.org/EIPS/eip-4524:
https://ethereum-magicians.org/t/why-isnt-there-an-erc-for-safetransfer-for-erc-20/7604/21

Reply:
We can extract it as a standalone library in the `utils` folder

Reply:
Ref:
https://github.com/valory-xyz/autonolas-registries/blob/main/contracts/ServiceRegistryTokenUtility.sol#L383
That is, our assumptions about the token must coincide with those we made earlier.

Reply:
ok